### PR TITLE
Adding GraphQL support for stat and listing queries

### PR DESCRIPTION
**Pull request by johnworth**
## Summary

This includes the type definitions, resolvers, and data sources needed for the queries. 

To try things out:
* Run `npm run dev`. 
* Hit `http://localhost:3000/login` in a browser. 
* Log in with your account. 
* Hit `http://localhost:3000/graphql` in the same browser to see the graphql UI. 

You'll have to modify the parameters in the query to work with your account.

## References:
* https://graphql.org/learn/schema/#interfaces
* https://graphql.org/learn/schema/#union-types
* https://www.apollographql.com/docs/apollo-server/schema/unions-interfaces/

### stat query:
```
{
  stat: filesystemStat(paths: [
    "/iplant/home/ipcdev/testing-vice", 
    "/iplant/home/ipcdev/testing-vice/drop-notes.md"
  ]) {
    id
    name: label
    path
    type
    
    ... on File {
      fileSize
    }
    
    ... on Folder {
      fileCount
    }
  }
}
```
The `... on File` and `... on Folder` blocks define which fields we want to grab for each type, since the result is a list of both types mixed together. The fields above that (id, label, path, and type) are common to the File and Folder types so they don't need to be split out the way that the fileSize and fileCount fields are. 

I renamed the label field to `name` just to show how that can be controlled in a query.

There are more fields available than what are listed here, check out the schema in the graphql UI for more.

#### Result
```
{
  "data": {
    "stat": [
      {
        "id": "3a33976c-5567-11e9-80be-d8d385e427d4",
        "name": "testing-vice",
        "path": "/iplant/home/ipcdev/testing-vice",
        "type": "DIR",
        "fileCount": 4
      },
      {
        "id": "2dc7d5d8-5567-11e9-80be-d8d385e427d4",
        "name": "drop-notes.md",
        "path": "/iplant/home/ipcdev/testing-vice/drop-notes.md",
        "type": "FILE",
        "fileSize": 856
      }
    ]
  }
}
```

### Folder listing:
```
{
  folderListing: listFolder(path: "/iplant/home/ipcdev") {
    stat {
      id
      name: label
      path
      permission
      type
    }
    listing {
      id
      name: label
      path
      permission
      type
    }
  }
}
```
I nested the stat information so that we don't have to create a duplicate type containing all of the same info as the Folder type but with the added listing field. This way we can define a FolderListing as :

```
type FolderListing {
    stat: Folder
    listing: [FilesystemObject]
}
```

The `type` field always returns upper-case, though the entity-type parameter should accept lower-case values. Not exactly sure how to fix that without losing specificity in the types (I'm using an enum to define the range of possible types).

#### Result
```
{
  "data": {
    "folderListing": {
      "stat": {
        "id": "e23e208c-e210-11e3-b8e3-6abdce5a08d5",
        "name": "ipcdev",
        "path": "/iplant/home/ipcdev",
        "permission": "OWN",
        "type": "FOLDER"
      },
      "listing": [
        {
          "id": "4378cca2-e212-11e3-b983-6abdce5a08d5",
          "name": "CORE-5257",
          "path": "/iplant/home/ipcdev/CORE-5257",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "e29e4e12-e210-11e3-b8e3-6abdce5a08d5",
          "name": "CORE-5413",
          "path": "/iplant/home/ipcdev/CORE-5413",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "850d5c64-e212-11e3-b993-6abdce5a08d5",
          "name": "Sub",
          "path": "/iplant/home/ipcdev/Sub",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "851497a4-e212-11e3-b993-6abdce5a08d5",
          "name": "TestCreateFolder",
          "path": "/iplant/home/ipcdev/TestCreateFolder",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "63ab1806-f62f-11e7-bf82-d8d385e427d4",
          "name": "TestData_dpt",
          "path": "/iplant/home/ipcdev/TestData_dpt",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "e2773be2-e210-11e3-b8e3-6abdce5a08d5",
          "name": "analyses",
          "path": "/iplant/home/ipcdev/analyses",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "59ca58ec-f62f-11e7-bf82-d8d385e427d4",
          "name": "analyses_dpt",
          "path": "/iplant/home/ipcdev/analyses_dpt",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "80f57eca-0d1c-11e8-803c-d8d385e427d4",
          "name": "analyses_qa",
          "path": "/iplant/home/ipcdev/analyses_qa",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "e26a5dbe-e210-11e3-b8e3-6abdce5a08d5",
          "name": "analyses_qa-1",
          "path": "/iplant/home/ipcdev/analyses_qa-1",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "b7d737ac-7f70-11e6-bf34-d8d385e427d4",
          "name": "analyses_qa-3",
          "path": "/iplant/home/ipcdev/analyses_qa-3",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "a9db662e-ed26-11e4-ac76-3c4a92e4a804",
          "name": "coge_data",
          "path": "/iplant/home/ipcdev/coge_data",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "239dc8f2-9693-11e9-80c6-d8d385e427d4",
          "name": "hello-shiny",
          "path": "/iplant/home/ipcdev/hello-shiny",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "e8b9468a-0bd1-11e5-8297-3c4a92e4a804",
          "name": "logs",
          "path": "/iplant/home/ipcdev/logs",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "72aa19a4-d01e-11e9-80d5-d8d385e427d4",
          "name": "lol",
          "path": "/iplant/home/ipcdev/lol",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "a2327096-71c7-11e4-904d-3c4a92e4a804",
          "name": "nex",
          "path": "/iplant/home/ipcdev/nex",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "3c1aeb2e-920c-11e5-81de-3c4a92e4a804",
          "name": "test",
          "path": "/iplant/home/ipcdev/test",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "989f44ee-720e-11e5-87bf-3c4a92e4a804",
          "name": "test-bulk-metadata",
          "path": "/iplant/home/ipcdev/test-bulk-metadata",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "c38e9696-7bcf-11e8-8071-d8d385e427d4",
          "name": "test-shiny-app",
          "path": "/iplant/home/ipcdev/test-shiny-app",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "3a33976c-5567-11e9-80be-d8d385e427d4",
          "name": "testing-vice",
          "path": "/iplant/home/ipcdev/testing-vice",
          "permission": "OWN",
          "type": "FOLDER"
        },
        {
          "id": "70de0646-11cc-11e4-ba8f-3c4a92e4a804",
          "name": "4store-1.1.5.dmg",
          "path": "/iplant/home/ipcdev/4store-1.1.5.dmg",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "07ffbe2c-ed92-11e6-bf47-d8d385e427d4",
          "name": "CORE-8424.txt",
          "path": "/iplant/home/ipcdev/CORE-8424.txt",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "3f64162e-eab5-11e4-8068-3c4a92e4a804",
          "name": "PDAP.fel.tree",
          "path": "/iplant/home/ipcdev/PDAP.fel.tree",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "b40be208-7bdb-11e8-8071-d8d385e427d4",
          "name": "demo-file.txt",
          "path": "/iplant/home/ipcdev/demo-file.txt",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "919e444c-9af8-11e6-bf3b-d8d385e427d4",
          "name": "path.list",
          "path": "/iplant/home/ipcdev/path.list",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "a4a681ca-6072-11e4-9d04-3c4a92e4a804",
          "name": "sample.csv",
          "path": "/iplant/home/ipcdev/sample.csv",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "8f9119d4-f62f-11e7-bf82-d8d385e427d4",
          "name": "sample1.newick",
          "path": "/iplant/home/ipcdev/sample1.newick",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "4a24a3fc-54bb-11e4-9ce0-3c4a92e4a804",
          "name": "savedParameters.txt",
          "path": "/iplant/home/ipcdev/savedParameters.txt",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "d43dff62-cae1-11e7-bf75-d8d385e427d4",
          "name": "screenshot-localhost 9091-2017-05-03-08-36-13.png",
          "path": "/iplant/home/ipcdev/screenshot-localhost 9091-2017-05-03-08-36-13.png",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "7617ea14-54bb-11e4-8cb6-3c4a92e4a804",
          "name": "soemthiong",
          "path": "/iplant/home/ipcdev/soemthiong",
          "permission": "OWN",
          "type": "FILE"
        },
        {
          "id": "43c93a70-e212-11e3-b983-6abdce5a08d5",
          "name": "tmp.txt",
          "path": "/iplant/home/ipcdev/tmp.txt",
          "permission": "OWN",
          "type": "FILE"
        }
      ]
    }
  }
}
```

More testing needs to be done with this, but it's a start at least and the basic functionality should work.

### CODE REVIEW

**Comment by ianmcorvidae** [comment]
Some general comments just from playing around in the interface:

a.) the graphql playground interface is super nice in general
b.) does it have some nice way to look at the model definitions right in it? I couldn't find one
c.) looks like we have to uppercase for the entityType parameter too, but that's not a huge burden or anything
d.) It looks like the stats that come from the listing are a little bit undone, particularly, not all fields are available. I'm guessing this is basically the thing you were talking about in chat the other day and it's a matter of figuring out what to do with it (e.g., I added fileCount to the `stat` section of a listing, and it came up null). I guess we'll either need to add some of that stuff to the listing endpoints or make stat queries.
e.) I notice one missing stat key, which is 'content-type' for files.
f.) we should definitely expose the data-info `path-info` endpoint in terrain and use it here for `stat`, because it's the same as stat except it allows specifically filtering to which fields are included (and can provide significant performance benefits depending what's being requested). That doesn't go in this PR though, just a general note

**Comment by ianmcorvidae** [comment]
(just for the sake of keeping track, missing stuff in listing that can be gotten from stat: `shareCount`, `fileCount`, `md5`, and presumably content type but that's supported nowhere in this code right now as mentioned)

**Comment by ianmcorvidae** [comment]
Okay, one more question/thought: I don't really like the structure of the models. I'd really like it if these could be combined and restructured into a different structure, where a query like this would work:

```
{ folder(path: "/iplant/home/ipcdev") {
    id
    path
    children(entityType: FILE) {
        id
        path
    }
}
```

(I'm not really picky about the field names, particularly `children`, which might not be the right thing there. So handwave over that right now. It could even make sense to have `subfolders` and `subfiles` as separate fields, or something like that.)

And, presumably, also for the top level stuff like `file`, as well as (I guess?) `files` and `folders` for the multiple-types-of-things requests.

Basically, I'd like it if we can make our queries describe _data_ and not describe _operations_ wherever possible, and `stat` and `folderListing` are definitely operations in my mind. But I have no idea how sensible/possible this is.

**Comment by ianmcorvidae** [comment]
heh, I lied about being done. in continuation of the last comment, I wonder if `filesystem` as the top-level unified query type would be best, and have that always return an array of filesystem objects. That way we don't need to separately implement `folders`/`folder`/`files`/`file` or anything like that.

**Comment by johnworth** [comment]
The goal here is to get authed queries working with the first draft of one of the more annoying parts of our data model. Probably should have stated that up front. I fully expect to have to iterate on the data model a LOT, but I want to get the UI plumbing work and first draft of the logic (the important part being the service calls) into master so other people can contribute directly more easily.

I tried having the listing as part of the Folder type multiple times (it's why it took me so long to get this up), but I concluded that there wasn't a good way of doing it without making multiple expensive service calls. To embed the listing in the Folder means that we should really support any combination of fields including the listing (e.g. Folder.listing and Folder.permission together in a query). My understanding is that that requirement means that I would have to have a separate resolver for each field in the Folder type and that Apollo would then hammer the crap out of Terrain for each additional field included in a query. A work around would be a caching dataloader (Facebook open-sourced one), but I'm trying to keep the scope of this pull request in check. Even with a dataloader, we'd have to make at least two calls, since we don't want to have to grab a listing along with basic stat info if it's not being used in the query.

**Comment by johnworth** [comment]
Also, some of the jankery with the GraphQL data model comes down to jankery we had to do in the API to keeps calls efficient. Changing the Terrain API a ton to make it more usable with GraphQL is way outside the scope of this PR.

File: src/server/graphql/dataSources/TerrainDataSource.js [hunk 1]
```diff
@@ -21,4 +35,76 @@ export default class TerrainDataSource extends RESTDataSource {
...
+
+    /**
+     * Returns a folder listing.
+     * @param {string} path - The full path to the folder being listed.
+     * @param {number} limit - The maximum number of items included in the response.
+     * @param {number} offset - The number of items to skip over before starting the list.
+     * @param {string} entityType - The type of items included in the results. Should be one of FILE, FOLDER, or ANY.
+     * @param {string} sortColumn - The column to sort on. One of NAME, ID, LASTMODIFIED, DATECREATED, SIZE, or PATH.
+     * @param {string} sortDirection - One of "ASC" (for ascending order) or "DESC" (for descending order).
+     * @returns {Array<{(File | Folder)}>}
```
[the remark this answers is not in the document]
Comment (psarando) [reply src/server/graphql/dataSources/TerrainDataSource.js]: Looks like this function is returning an array in #17, so ignore this comment now.

File: src/server/graphql/typesDefs/Filesystem.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ * @property {string} id - The UUID assigned to the folder.
+ * @property {string} label - The label for the folder. Usually the folder name without the path.
+ * @property {string} path - Full path to the folder in the data store.
+ * @property {string} permission - The maximum permission level on the folder for the user. One of "read", "write", or "own".
+ * @property {number} dateCreated - Folder creation date as milliseconds since the epoch.
+ * @property {number} dateModified - Last modified date as milliseconds since the epoch.
+ * @property {string} type - Either "file" or "folder". Should be "folder".
+ * @property {number} shareCount - The number of users the folder has been shared with.
+ * @property {number} fileCount - The number of files contained in the folder.
+ * @property {Array<Object>} listing - A list of files and folders contained in the folder.
```
[the remark this answers is not in the document]
Comment (psarando) [reply src/server/graphql/typesDefs/Filesystem.js]: Oh wait, I think IntelliJ is tricking me, and both the `dir-count` and `file-count` fields for folders are used in the details panel in the current UI.

[the remark this answers is not in the document]
Comment (johnworth) [reply src/server/graphql/typesDefs/Filesystem.js]: I think that some of these are going to have to get added to the return value from the stat endpoint. Namely, `badName`, `hasSubDirs`, and `isFavorite`. They're currently only presented through the the paged-directory endpoint, which makes them hard to get at for non-listing purposes.

`files` and `folders` have been replaced by the `contents` field of the Folder type.

Unless I'm missing something, I think `total` is redundant since both `fileCount` and `folderCount` are included, so total can be computed. That's better done in the UI logic since computing it in Apollo means introducing a dependency between two fields, which is frowned upon and painful to implement in resolvers (AFAIK). 

Not sure how to handle `totalBad` (which sounds like early 90's slang to me).

File: src/server/graphql/dataSources/TerrainDataSource.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnworth) [reply src/server/graphql/dataSources/TerrainDataSource.js]: Thanks for catching this, I have to update the JSDoc in general. Forgot about it with all of the changes flying around.

File: src/server/graphql/dataSources/TerrainDataSource.js [hunk 3]
```diff
@@ -21,4 +35,76 @@ export default class TerrainDataSource extends RESTDataSource {
...
+                return f;
+            })
+        );
+
+        retval.listing = retval.listing.concat(
+            _.map(responseValue.files, (f) => {
+                f.type = FILE;
+                return f;
+            })
+        );
```
[the remark this answers is not in the document]
Comment (johnworth) [reply src/server/graphql/dataSources/TerrainDataSource.js]: Thanks, I'll get this change in. I got a new tool (lodash) and suddenly wanted to use it everywhere.

File: src/server/graphql/typesDefs/Filesystem.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (psarando) [reply src/server/graphql/typesDefs/Filesystem.js]: > I think that some of these are going to have to get added to the return value from the stat endpoint. Namely, `badName`, `hasSubDirs`, and `isFavorite`. They're currently only presented through the the paged-directory endpoint, which makes them hard to get at for non-listing purposes.
> 

Good point.

> Unless I'm missing something, I think `total` is redundant since both `fileCount` and `folderCount` are included, so total can be computed. That's better done in the UI logic since computing it in Apollo means introducing a dependency between two fields, which is frowned upon and painful to implement in resolvers (AFAIK).
> 

The `total` field is specific to the listing endpoint and is required for paging in the UI. It can also vary if the query is doing any filtering (just return files I own), although I don't think we currently support filtering.

> Not sure how to handle `totalBad` (which sounds like early 90's slang to me).

🤣
I'm can't remember how this is getting used currently, but it's related to the `badName` flags on any of the folder's contents. I'm not sure if we'll need that in the new UI, so it can be added later if we need it later.

File: src/server/graphql/dataSources/TerrainDataSource.js [hunk 4]
```diff
@@ -21,4 +35,76 @@ export default class TerrainDataSource extends RESTDataSource {
...
+    /**
+     * Returns info about a file or folder from terrain.
+     * @param {string} fullPath - The full path to the file in the data store.
+     * @returns {(File | Folder)}
+     */
+    async filesystemStat(fullPaths) {
+        let retval = await this.post("/secured/filesystem/stat", {
+            paths: fullPaths,
+        });
+        return _.values(retval.paths);
```
[the remark this answers is not in the document]
Comment (johnworth) [reply src/server/graphql/dataSources/TerrainDataSource.js]: This was fixed when I switched over to dataloader earlier, I believe.
